patches/tModLoader/Terraria/ModLoader/ChestLoot.Entry.cs

```diff
@@ -0,0 +1,96 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System;
+using Terraria.ID;
+
+namespace Terraria.ModLoader;
+
+public partial class ChestLoot {
+	private static Item EmptyInstance => new();
+
+	public class Entry {
+		internal readonly Item item;
+		private readonly List<ICondition> conditions;
+		private readonly bool askingNicelyToNotAdd = false;
+		private Ref<bool> hide;
+
+		public Item Item => item;
+
+		public Dictionary<bool, List<Entry>> ChainedEntries;
+
+		public Entry(params ICondition[] condition) : this(EmptyInstance, condition) {
+			askingNicelyToNotAdd = true;
+		}
+
+		public Entry(int item, params ICondition[] condition) : this(ContentSamples.ItemsByType[item], condition) { }
+
+		public Entry(Item item, params ICondition[] condition) {
+			hide = new(false);
+			this.item = item;
+			conditions = condition.ToList();
+
+			ChainedEntries = new()
+			{
+				{ false, new() },
+				{ true, new() }
+			};
+		}
+
+		public Entry OnSuccess(int itemId, params ICondition[] condition) {
+			return OnSuccess(new Entry(ContentSamples.ItemsByType[itemId], condition));
+		}
+
+		public Entry OnSuccess(Entry entry) {
+			ChainedEntries[true].Add(entry);
+			return this;
+		}
+
+		public Entry OnFail(int itemId, params ICondition[] condition) {
+			return OnFail(new Entry(ContentSamples.ItemsByType[itemId], condition));
+		}
```

Comment:
Is `OnFail` necessary? That seems like it would be more appropriate for things that were chance based.
It seems like it would be appropriate for, well, `ChestLoot` or full loot tables, but since this is just conditional visibility I think it might be better to have all conditions be 'top level' (like recipes). Needs discussion